### PR TITLE
Add transitive reduction algorithm

### PR DESCRIPTION
**Pull request by nrkramer**
*A good pull request has the following information:*
- Addresses https://github.com/ZigRazor/CXXGraph/issues/273
- Adds the transitive reduction algorithm and supporting test
- Adds a function for finding an edge within the graph given two nodes
- @mentions of the person or team responsible for reviewing proposed changes. ( optional ) 
  - @nrkramer

### CODE REVIEW

File: include/Graph/Graph.hpp [hunk 1]
```diff
@@ -671,9 +714,9 @@ const std::set<const Node<T> *> Graph<T>::getNodeSet() const {
 template <typename T>
 void Graph<T>::setNodeData(const std::string &nodeUserId, T data) {
   auto nodeset = this->nodeSet();
-  auto nodeIt = std::find_if(nodeset.begin(), nodeset.end(), [&nodeUserId](auto node){
-		return node->getUserId() == nodeUserId;
-	  });
+  auto nodeIt = std::find_if(
+      nodeset.begin(), nodeset.end(),
+      [&nodeUserId](auto node) { return node->getUserId() == nodeUserId; });
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:719]: Automatically done by clang-format

Comment (nrkramer) [inline include/Graph/Graph.hpp:719]: Automatically done by clang-format

File: include/Graph/Graph.hpp [hunk 2]
```diff
@@ -2910,9 +2985,9 @@ int Graph<T>::writeToMTXFile(const std::string &workingDir,
   std::string header = "%%MatrixMarket graph";
   // Check if the adjacency matrix is symmetric, i.e., if all the edges are
   // undirected
-  bool symmetric = !std::any_of(edgeSet.begin(), edgeSet.end(), [](auto edge){
-		return (edge->isDirected().has_value() && edge->isDirected().value());
-	  });
+  bool symmetric = !std::any_of(edgeSet.begin(), edgeSet.end(), [](auto edge) {
+    return (edge->isDirected().has_value() && edge->isDirected().value());
+  });
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:2990]: Automatically done by clang-format

File: include/Graph/Graph.hpp [hunk 3]
```diff
@@ -2999,7 +3074,7 @@ int Graph<T>::readFromMTXFile(const std::string &workingDir,
 
   // Since the matrix represents the adjacency matrix, it must be square
   if (n_rows != n_cols) {
-	return -1;
+    return -1;
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:3077]: Automatically done by clang-format

File: include/Graph/Graph.hpp [hunk 4]
```diff
@@ -3029,8 +3104,9 @@ int Graph<T>::readFromMTXFile(const std::string &workingDir,
   }
 
   if (n_edges != edgeMap.size()) {
-	std::cout << "Error: The number of edges does not match the value provided in the size line.\n";
-	return -1;
+    std::cout << "Error: The number of edges does not match the value provided "
+                 "in the size line.\n";
+    return -1;
```
Comment (nrkramer) [inline include/Graph/Graph.hpp:3109]: Automatically done by clang-format

**Review by ZigRazor — CHANGES_REQUESTED** [review]
Do you can check if is possible to eliminate at least the "violations" reported by cppcheck?

File: include/Graph/Graph.hpp (hunk 1 above)
Comment (ZigRazor) [reply include/Graph/Graph.hpp:719]: ok

File: include/Graph/Graph.hpp (hunk 3 above)
Comment (ZigRazor) [reply include/Graph/Graph.hpp:3077]: ok

File: include/Graph/Graph.hpp (hunk 4 above)
Comment (ZigRazor) [reply include/Graph/Graph.hpp:3109]: ok

File: include/Graph/Graph.hpp (hunk 2 above)
Comment (ZigRazor) [reply include/Graph/Graph.hpp:2990]: ok

**Comment by nrkramer** [comment]
> Do you can check if is possible to eliminate at least the "violations" reported by cppcheck?

I've fixed a few of the "violations". Some of them seem intended for C, such as MISRA 17.8, as reference parameter modification is bread-and-butter in C++ (in C++, programmers use `const` to denote that the reference should not be modified in the function body).

Some of the other violations, such as MISRA 15.5, are necessary as optimizations. Early return from function is common practice to speed up function calls, and refractoring the function to support leaving loops quickly would result in code that is difficult to understand.

I've dismissed the violations on code in which this PR is not concerned, as it is irrelevant to the PR.

**Review by ZigRazor — APPROVED** [review]
Good Job!

**Pull request merged**